### ISSUE TITLE
Improvement of actual timings for test body without call overhead?

### ISSUE THREAD

**Issue by Poimen** [nickbruun/hayai#25]
Hi,

I have been looking over the framework - but not actually tried any thing with it yet. Hence, if the below does not make sense, then you may close the issue.

From inspection it looks like the TestBody() timing includes overhead for both calls to the clock and the loop consideration.

If I look at the code in [hayai_test.hpp:49](https://github.com/nickbruun/hayai/blob/master/src/hayai_test.hpp#L49):
```
startTime = Clock::Now();

// Run the test body for each iteration.
std::size_t iteration = iterations;
while (iteration--)
   TestBody();

// Get the ending time.
endTime = Clock::Now();
````
endTime - startTime  would also represent the time taken to return from the functions Clock::Now() , processing of ```iteration--``` and the function call to ```TestBody```.

Would it not be better to:

1. Bench Clock::Now() to find the function call hit for it, then subtract that from the time
2. Use ```--iteration``` (prefix is generally faster), but also bench the time taken (probably very small, so might be too negligible) to do the subtraction and the branch consideration (i.e. the ```while``` condition)
3. Bench a "dummy" function call for ```TestBody()``` that can negate the call time over a virtual function call

In my mind, currently:
```Test Time = Time in body + Time in framework```

While ```Time in framework``` may stay relatively static, there is no way to know how much time is in the framework so exact timings of the ```TestBody()``` call is not known.

If you take consideration of the 1,2 and 3 points, the time can be removed from the final calculation to formulate a more accurate timing.

Would this be worthwhile?

Regards,
Richard

**Comment by nickbruun** [comment]
Hi Ricard,

You're absolutely right in this observation. I'll make this part of the push for 1.0 :) :+1:

Nick

**Comment by nickbruun** [comment]
Implemented in #31.
